### DISCUSSION TITLE
Duplicate Material property into AuxVariable

### DISCUSSION THREAD

**Discussion by hokonkim** [idaholab/moose#21104] [Q&A General] [answered]
Hello, I have a question about duplicating material property into AuxVariable.

I tried `MaterialRealAux`. But this outputs element volume-averaged material properties.

I just want to capture the pure value of material properties.

The reason why I duplicate material property into AuxVariable is "Material property doesn't have the xyz coordinate, but AuxVariable have the xyz coordinate after duplication."

I want to track the change in Material properties matching xyz coordinates.

What will be the best solution for me?

**Comment by GiudGiud** [comment]
Hello

If the auxvariable is a nodal variable you ll get a nodal value not a volume averaged one. 
What do you mean by 'pure' value?

What do you mean by 'doesn't have the xyz coordinate' as well?

Guillaume

**Reply by hokonkim** [reply]
[Materials]
      [./tpbActivity]
        type  = ParsedMaterial
        block = 'PT_TPB_TET4'
        f_name = 'tpbActivity_S0'
        function = 'if(time = 0,
                            s0_tpb_init,
                            if(((V_Em - VolCr) / V_Em)>=0, s0_tpb_init * ((V_Em - VolCr) / V_Em), 0))'
        args = 'V_Em s0_tpb_init'
        material_property_names = 'VolCr time'
      [../]
    []

    [AuxVariables]
      [./Copy_tpbActivity_S0]
        family = MONOMIAL
        block = 'PT_TPB_TET4'
      [../]
    []

    [AuxKernels]
      [./copy_tpbActivity]
        type = MaterialRealAux
        property = tpbActivity_S0
        variable = Copy_tpbActivity_S0
        block = 'PT_TPB_TET4'
      [../]
    []

Above is a part of my code.

I used "Pure" for the value without any additional processes(e.g., volume-averaged).

If I export the Material properties(Cell data in Paraview) through Paraview, then it didn't have XYZ coordinates.

But variables (Point data in Paraview) have XYZ coordinates.

**Reply by GiudGiud** [reply]
how do you export the material properties?

Guillaume

**Reply by hokonkim** [reply]
If I add command "output = exodus" in Material properties, then in the result file(*.e), the material property is stored as Cell data(in paraview).

Then, I can export (save) the material property.
<img width="1031" alt="Screen Shot 2022-05-23 at 6 48 46 PM" src="https://user-images.githubusercontent.com/46009457/169916981-b3df3bb0-2402-473b-a4d0-a06d3980c6a8.png">

**Reply by GiudGiud** [reply]
ok so the the default output from the materials block does this.
But with a nodal aux variable and MaterialRealAux you can get the nodal values. then it ll be point data

**Reply by hokonkim** [reply]
How can I make a nodal aux variable from the current situation??

**Reply by GiudGiud** [reply]
```
[AuxVariables]
  [some_name]
  []
[]

[AuxKernels]
  [convert]
    type = MaterialRealAux
    variable = some_name
    property = 'name of the property'
  []
[]
```

**Reply by hokonkim** [reply]
As far as I remember, without specifying “family = monomial” in AuxVariables, I could not run the simulation properly. 

I will try the way as you suggested.

**Reply by GiudGiud** [reply]
we are talking about output here so you should be able to use the lagrange family

**Reply by hokonkim** [reply]
I got the error like below
<img width="849" alt="Screen Shot 2022-05-23 at 8 54 16 PM" src="https://user-images.githubusercontent.com/46009457/169927124-916e2e8c-a624-4ef9-9fdf-044a5ee111d5.png">

**Reply by GiudGiud** [reply]
umm ok. This is because it would be undefinied at a subdomain boundary, like we would not know which expression to use to compute the material property if it was different on both sides of the node

Can you use AuxVariables with CONSTANT order MONOMIAL family to save the X, Y and Z coordinates then use that as the cell data?
You'd just need a FunctionAux and ParsedFunction to do this

**Reply by hokonkim** [reply]
[Functions]
      [./func_coord_X]
        type = ParsedFunction
        value = 'x'
      [../]

      [./func_coord_Y]
        type = ParsedFunction
        value = 'y'
      [../]

      [./func_coord_Z]
        type = ParsedFunction
        value = 'z'
      [../]
    []

    [AuxVariables]
      [./coord_X]
        order = CONSTANT
        family = MONOMIAL
        block = 'PT_TPB_TET4'
      [../]

      [./coord_Y]
        order = CONSTANT
        family = MONOMIAL
        block = 'PT_TPB_TET4'
      [../]

      [./coord_Z]
        order = CONSTANT
        family = MONOMIAL
        block = 'PT_TPB_TET4'
      [../]
    []

    [AuxKernels]
      [./Coordinates_X]
        type = FunctionAux
        variable = coord_X
        function = func_coord_X
      [../]

      [./Coordinates_Y]
        type = FunctionAux
        variable = coord_Y
        function = func_coord_Y
      [../]

      [./Coordinates_Z]
        type = FunctionAux
        variable = coord_Z
        function = func_coord_Z
      [../]
    []

When I used the "only one coordinate(X)", I can get the result as what I expected.

But, when add two more coordinates, I cannot get XYZ coordinates properly.

**Reply by GiudGiud** [reply]
Where did you add two coordinates? In which field?

Also you can use vector variables and a VectorFunctionAux
https://mooseframework.inl.gov/source/auxkernels/VectorFunctionAux.html

**Reply by hokonkim** [reply]
After I renamed the "X" "Y" "Z" into "X1" "Y1" "Z1". I can get the result like this.


<img width="672" alt="Screen Shot 2022-05-24 at 12 21 38 AM" src="https://user-images.githubusercontent.com/46009457/169948323-ff156c0b-f778-49dd-9fca-bed7e8f7c805.png">


~~By the way, I think X1, Y1, and Z1 doesn't represent XYZ coordinates properly. Only Z1 looks like worked?~~

This XYZ coordinates were applied at only "PT_TPB_TET4" block.

I think it is right.

**Comment by ngrilli** [comment, marked as answer]
Dear @hokonkim 

I think I do something similar here:
https://github.com/ngrilli/c_pfor_am/blob/main/test/tests/SlipGradients/polycrystal_initial_GND.i

I define first monomial auxvariables, e.g. slip_increment_1,
I declare and assign a material property in the material object CrystalPlasticityDislocationUpdate,
then I use MaterialStdVectorAux to assign the material property to the AuxVariable.

This provide a linearly interpolated material property over the elements
in the form of an AuxVariable.

Hope this helps,
Best Regards,
Nicolò

**Reply by Vandenbg** [reply]
Hello @ngrilli,

I can't seem to find the syntax page for CrystalPlasticityDislocationUpdate, could you point it out to me?

Thank you,
Gavin

**Reply by GiudGiud** [reply]
is it any of those?
https://mooseframework.inl.gov/source/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdate.html
https://mooseframework.inl.gov/source/materials/crystal_plasticity/CrystalPlasticityTwinningKalidindiUpdate.html

**Reply by Vandenbg** [reply]
It looks like the name may have changed at some point, that is the source of my confusion. Thank you!
